cupyx/fallback_mode/ndarray.py

```diff
@@ -0,0 +1,130 @@
+"""
+class ndarray is wrapper around cupy.ndarray
+to support fallback of methods of type `ndarray.func()`
+"""
+import numpy as np
+
+import cupy as cp
+from cupyx.fallback_mode import data_transfer
+
+try:
+    from cupyx.fallback_mode import fallback
+except ImportError:
+    import sys
+    fallback = sys.modules[__package__ + '.fallback']
+
+
+class ndarray:
+    """
+    Wrapper around cupy.ndarray
+    Gets initialized with a cupy ndarray.
+    """
+
+    def __init__(self, array):
+        self._array = array
+
+    def __getattr__(self, attr):
+        """
+        Catches attributes corresponding to ndarray.
+
+        Args:
+            attr (str): Attribute of ndarray class.
+
+        Returns:
+            (_RecursiveAttr object, self._array.attr):
+            Returns_RecursiveAttr object with numpy_object, cupy_object.
+            Returns self._array.attr if attr is not callable.
+        """
+
+        cupy_object = getattr(cp.ndarray, attr, None)
+
+        numpy_object = getattr(np.ndarray, attr)
+
+        if not callable(numpy_object):
+            return getattr(self._array, attr)
+
+        return fallback._RecursiveAttr(numpy_object, cupy_object, self)
+
+    def _get_array(self):
+        """
+        Returns _array (cupy.ndarray) of ndarray object.
+        """
+        return self._array
+
+    @property
+    def __class__(self):
+        """
+        Redirects __class__ of ndarray object to cupy.ndarray
+        Used while execution of isinstance() method.
+        """
+        return cp.ndarray
+
+
+def _get_cupy_ndarray(args, kwargs):
+    return data_transfer._get_xp_args(
+        ndarray, ndarray._get_array, (args, kwargs))
+
+
+def _get_fallback_ndarray(cupy_res):
+    return data_transfer._get_xp_args(cp.ndarray, ndarray, cupy_res)
+
+
+# Decorator for ndarray magic methods
+def make_method(name):
+    def method(self, *args, **kwargs):
+
+        args, kwargs = _get_cupy_ndarray(args, kwargs)
+
+        cupy_method = getattr(cp.ndarray, name)
+
+        res = cupy_method(self._array, *args, **kwargs)
+
+        return _get_fallback_ndarray(res)
+
+    return method
```

Comment:
Generated methods have no docstrings. Use `functools.update_wrapper`.

Reply:
This will work for `fallback_mode.ndarray.__add__.__doc__` not `numpy.ndarray.__add__.__doc__`, since  later is an instance of `_RecursiveAttr`.

Reply:
We can add following to `_RecursiveAttr`:
```python
@property
def __doc__(self):

    if self._cupy_object is not None:
        return self._cupy_object.__doc__

    return self._numpy_object.__doc__
```

Reply:
It seems OK to just return NumPy's docs.

Reply:
Done